system/loggerd/deleter.py

```diff
@@ -5,22 +5,43 @@
 from system.swaglog import cloudlog
 from system.loggerd.config import ROOT, get_available_bytes, get_available_percent
 from system.loggerd.uploader import listdir_by_creation
+from system.loggerd.xattr_cache import getxattr
 
 MIN_BYTES = 5 * 1024 * 1024 * 1024
 MIN_PERCENT = 10
 
 DELETE_LAST = ['boot', 'crash']
 
+PRESERVE_ATTR_NAME = 'user.preserve'
+PRESERVE_ATTR_VALUE = b'1'
+PRESERVE_COUNT = 5
+
+
+def has_preserve_xattr(d: str) -> bool:
+  return getxattr(os.path.join(ROOT, d), PRESERVE_ATTR_NAME) == PRESERVE_ATTR_VALUE
+
 
 def deleter_thread(exit_event):
   while not exit_event.is_set():
     out_of_bytes = get_available_bytes(default=MIN_BYTES + 1) < MIN_BYTES
     out_of_percent = get_available_percent(default=MIN_PERCENT + 1) < MIN_PERCENT
 
     if out_of_percent or out_of_bytes:
+      dirs = listdir_by_creation(ROOT)
+
+      # skip deleting most recent N preserved segments (and their prior segment)
+      preserved_dirs = []
+      for n, d in enumerate(filter(has_preserve_xattr, reversed(dirs))):
+        if n == PRESERVE_COUNT:
+          break
+        date_str, _, seg_num = d.rpartition("--")
+
+        # preserve segment and its prior
+        preserved_dirs.append(d)
+        preserved_dirs.append(f"{date_str}--{int(seg_num) - 1}")
```

Comment:
needs some exception handling. what if I set the preserve xattr on the boot dir?